packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts

```diff
@@ -1,25 +1,149 @@
 import { BigNumber } from '@0x/utils';
 import * as _ from 'lodash';
+import { OptimizerCapture, route, SerializedPath } from 'neon-router';
+import { performance } from 'perf_hooks';
 
 import { MarketOperation } from '../../types';
 
 import { DEFAULT_PATH_PENALTY_OPTS, Path, PathPenaltyOpts } from './path';
-import { ERC20BridgeSource, Fill } from './types';
+import { ERC20BridgeSource, Fill, FillData } from './types';
 
 // tslint:disable: prefer-for-of custom-no-magic-numbers completed-docs no-bitwise
 
 const RUN_LIMIT_DECAY_FACTOR = 0.5;
+const RUST_ROUTER_NUM_SAMPLES = 200;
+const SHOULD_USE_RUST_ROUTER = process.env.RUST_ROUTER === 'true';
+
+function findOptimalRustPath(input: BigNumber, allFills: Fill[][]): Path {
+    // Track sample id's to integers (required by rust router)
+    const sampleIdLookup: { [key: string]: number } = {};
+    let sampleIdCounter = 0;
+    const fillToSampleId = (s: { source: string; sourcePathId: string; index: number }): number => {
+        const key = `${s.source}-${s.sourcePathId}-${s.index}`;
+        if (sampleIdLookup[key]) {
+            return sampleIdLookup[key];
+        } else {
+            sampleIdLookup[key] = ++sampleIdCounter;
+            return sampleIdLookup[key];
+        }
+    };
+
+    const adjustedParsedFills = allFills.map(fills => {
+        const adjustedFills: Fill[] = [];
+        // Samples are turned into Fills
+        // Fills are dependent on their parent and have their parents information "subtracted" from them
+        // e.g a samples for [1,10,100] => [5,50,500] look like [1, 9, 91] => [5, 40, 400]
+        for (let i = 0; i < fills.length; i++) {
+            const parsedFill: Fill = { ...fills[i] };
+            if (parsedFill.index !== 0) {
+                const parent = adjustedFills[i - 1];
+                parsedFill.parent = parent;
+                parsedFill.input = parsedFill.input.plus(parent.input);
+                parsedFill.output = parsedFill.output.plus(parent.output);
+                // Adjusted output is only modified for the first fill
+                const feePerc = parent.adjustedOutput.dividedBy(parent.output);
+                parsedFill.adjustedOutput = feePerc.times(parsedFill.output);
```

Comment:
FYI it occured to me when I was trying to sleep that this is wrong. 

The fee is a `fixed` cost and doesn't scale with the `output` so moving the output down by the % of previous is waaay over adjusting. It should be a constant for all the samples (e.g `fee = [3 DAI, ...3 DAI]`) and not a percentage  (e.g `fee = [3 DAI, 30 DAI, 300 DAI]`)

Reply:
Thanks 🙏  Hope you didn't lose too much sleep over it 😂